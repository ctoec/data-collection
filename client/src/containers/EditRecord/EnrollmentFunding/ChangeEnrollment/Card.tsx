import React, { useState, useEffect } from 'react';
import {
  Card,
  ExpandCard,
  Button,
  CardExpansion,
  Alert,
  CardProps,
} from '@ctoec/component-library';
import { Enrollment, Child } from '../../../../shared/models';
import { ChangeEnrollmentForm } from './Form';

type ChangeEnrollmentCardProps = CardProps & {
  child: Child;
  currentEnrollment?: Enrollment;
  key?: string;
  activeKey?: string;
  afterSaveSuccess: () => void;
};

/**
 * Component for gathering user input to change child's Enrollment.
 * Uses a ChangeEnrollment data object to enable the user to provide
 * enrollment end date and funding last reporting period for previously current data.
 */
export const ChangeEnrollmentCard: React.FC<ChangeEnrollmentCardProps> = ({
  child,
  currentEnrollment,
  afterSaveSuccess,
  expanded,
}) => {
  const [expandedCard, setExpandedCard] = useState(expanded);
  const [closeCard, setCloseCard] = useState(false);
  const [error, setError] = useState<string>();

  // Explicitly don't want `closeCard` as a dep, as this
  // needs to be triggered on render caused by child refetch
  // (not only when closeCard changes)
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    if (closeCard) setCloseCard(false);
  });

  return (
    <Card forceClose={closeCard}>
      <div className="display-flex flex-justify flex-row flex-align-center">
        {!currentEnrollment ? (
          <div className="usa-prose-body">
            {child.firstName} has no active enrollments
          </div>
        ) : (
          <div className="usa-prose-body">
            Has {child.firstName}'s age group and/or site changed?
          </div>
        )}
        <ExpandCard>
          <Button text="Change enrollment" appearance="outline" />
        </ExpandCard>
      </div>
      <CardExpansion>
        <h3 className="margin-top-2 margin-bottom-2">New enrollment</h3>
        {error && <Alert type="error" text={error} />}
        <ChangeEnrollmentForm
          afterSaveSuccess={() => {
            setCloseCard(true);
            afterSaveSuccess();
          }}
          afterSaveFailure={(err) => setError(err)}
          child={child}
          currentEnrollment={currentEnrollment}
        />
      </CardExpansion>
    </Card>
  );
};
