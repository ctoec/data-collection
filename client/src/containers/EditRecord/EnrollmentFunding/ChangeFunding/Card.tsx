import React, { useState } from 'react';
import { Card, Button } from '@ctoec/component-library';
import { Enrollment } from '../../../../shared/models';
import { ChangeFundingForm } from './Form';
import { RecordFormProps } from '../../../../components/Forms';
import { Heading, HeadingLevel } from '../../../../components/Heading';
import { getCurrentFunding } from '../../../../utils/models';

type ChangeFundingCardProps = {
  enrollment: Enrollment;
  orgId: number;
  afterSaveSuccess: () => void;
  setAlerts: RecordFormProps['setAlerts'];
  topHeadingLevel: HeadingLevel;
};

/**
 * Component for gathering user input to change enrollment's funding.
 * Uses a ChangeFunding data object to enable the user to provide
 * new funding data and/or last reporting period for previously current
 * funding, based on user's chosen action (start new funding vs end current funding)
 */
export const ChangeFundingCard: React.FC<ChangeFundingCardProps> = ({
  enrollment,
  orgId,
  afterSaveSuccess,
  setAlerts,
  topHeadingLevel,
}) => {
  const [visibleForm, setVisibleForm] = useState<'end' | 'start'>();
  const activeFunding = getCurrentFunding({ enrollment })
	
	return (
    <>
      {!!visibleForm && (
        <Card>
          <>
            <Heading level={topHeadingLevel}>
              {visibleForm === 'end' ? 'End current funding' : 'Change funding'}
            </Heading>
            <ChangeFundingForm
              afterSaveSuccess={() => {
                setVisibleForm(undefined);
                afterSaveSuccess();
              }}
              afterSaveFailure={(err) => {
                console.error(err);
                setAlerts([
                  {
                    type: 'error',
                    text:
                      'Unable to change funding. Make sure all required information is provided',
                  },
                ]);
              }}
              changeType={visibleForm}
              enrollment={enrollment}
              hideForm={() => setVisibleForm(undefined)}
              orgId={orgId}
            />
          </>
        </Card>
      )}
      <Card>
        <div className="display-flex flex-column flex-align-end">
          <div>
            {activeFunding && (
              <Button
                className="margin-right-4"
                text="End current funding"
                appearance="unstyled"
                onClick={() => setVisibleForm('end')}
              />
            )}
            <Button
              text={activeFunding ? 'Change funding' : 'Start new funding'}
              appearance="unstyled"
              onClick={() => setVisibleForm('start')}
            />
          </div>
        </div>
      </Card>
    </>
  );
};
