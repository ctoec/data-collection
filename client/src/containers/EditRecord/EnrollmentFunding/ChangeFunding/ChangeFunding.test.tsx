import React from 'react';
import {
  accessibilityTestHelper,
  snapshotTestHelper,
} from '../../../../testHelpers';
import { ChangeFundingCard } from './Card';
import {
  Child,
  Site,
  AgeGroup,
  FundingSource,
  ReportingPeriod,
  Funding,
} from '../../../../shared/models';
import moment from 'moment';
import { RenderResult } from '@testing-library/react';
import { fireEvent, wait } from '@testing-library/dom';

const enrollment = {
  child: {} as Child,
  id: 1,
  site: { id: 1, siteName: 'Site' } as Site,
  ageGroup: AgeGroup.InfantToddler,
  entry: moment.utc('2020-09-01'),
};

const enrollmentWithFunding = {
  ...enrollment,
  fundings: [
    {
      id: 1,
      fundingSpace: {
        ageGroup: AgeGroup.InfantToddler,
        source: FundingSource.CDC,
      },
      firstReportingPeriod: {} as ReportingPeriod,
    } as Funding,
  ],
};

const waitExpand = (buttonText: string) => async (
  renderResult: RenderResult
) => {
  const expandButton = await renderResult.findByText(buttonText);
  fireEvent.click(expandButton);
  await wait();
};

describe('EditRecord', () => {
  describe('EnrollmentFundng', () => {
    describe('ChangeFunding', () => {
      [
        {
          enrollment,
          before: waitExpand('Start new funding'),
          name: 'matches snapshot - without funding',
        },
        {
          enrollment: enrollmentWithFunding,
          before: waitExpand('Start new funding'),
          name: 'matches snapshot - start funding',
        },
        {
          enrollment: enrollmentWithFunding,
          before: waitExpand('End current funding'),
          name: 'matches snapshot - end funding',
        },
      ].forEach(({ enrollment, before, name }) => {
        snapshotTestHelper(
          <ChangeFundingCard
            enrollment={enrollment}
            orgId={1}
            afterSaveSuccess={jest.fn()}
            topHeadingLevel="h3"
          />,
          { before, name }
        );
        accessibilityTestHelper(
          <ChangeFundingCard
            enrollment={enrollment}
            orgId={1}
            afterSaveSuccess={jest.fn()}
            topHeadingLevel="h2"
          />,
          { before, name }
        );
      });
    });
  });
});
