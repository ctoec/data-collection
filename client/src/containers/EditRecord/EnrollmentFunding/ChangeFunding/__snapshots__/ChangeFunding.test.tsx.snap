// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditRecord EnrollmentFundng ChangeFunding matches snapshot - change funding 1`] = `
<DocumentFragment>
  <div
    class="oec-card"
  >
    <div
      class="oec-card-cell"
    >
      <h3>
        Change funding
      </h3>
      <form
        class="usa-form usa-form"
        id="change-funding-form"
      >
        <fieldset
          aria-labelledby="fieldset-legend-funding-source-1"
          class="usa-fieldset"
          id="funding-source-1"
        >
          <legend
            id="fieldset-legend-funding-source-1"
          >
            <span
              class="usa-label"
            >
              Funding source options
            </span>
          </legend>
          <div />
        </fieldset>
        <div
          class="usa-form-group"
        >
          <label
            class="usa-label"
            for="last-reporting-period"
          >
            Last reporting period for current CDC - Child Day Care - undefined funding
          </label>
          <span
            class="usa-hint text-italic"
          >
            Last reporting period is only required if this child's funding is no longer active.
          </span>
          <select
            aria-required="false"
            class="usa-select"
            id="last-reporting-period"
            name="last-reporting-period"
          >
            <option
              value=""
            >
              - Select -
            </option>
          </select>
        </div>
        <div
          class="display-flex flex-align-center"
        >
          <button
            class="usa-button usa-button--outline bx--btn bx--btn--primary"
            tabindex="0"
            type="button"
          >
            Cancel
          </button>
          <div
            class="margin-top-1"
          >
            <button
              class="usa-button bx--btn bx--btn--primary"
              tabindex="0"
              type="submit"
            >
              Change current funding
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      class=""
    />
  </div>
  <div
    class="oec-card"
  >
    <div
      class="oec-card-cell"
    >
      <div
        class="display-flex flex-column flex-align-end"
      >
        <div>
          <button
            class="usa-button usa-button--unstyled margin-right-4 bx--btn bx--btn--primary"
            tabindex="0"
            type="button"
          >
            End current funding
          </button>
          <button
            class="usa-button usa-button--unstyled bx--btn bx--btn--primary"
            tabindex="0"
            type="button"
          >
            Change funding
          </button>
        </div>
      </div>
    </div>
    <div
      class=""
    />
  </div>
</DocumentFragment>
`;

exports[`EditRecord EnrollmentFundng ChangeFunding matches snapshot - end funding 1`] = `
<DocumentFragment>
  <div
    class="oec-card"
  >
    <div
      class="oec-card-cell"
    >
      <h3>
        End current funding
      </h3>
      <form
        class="usa-form usa-form"
        id="change-funding-form"
      >
        <div
          class="usa-form-group"
        >
          <label
            class="usa-label"
            for="last-reporting-period"
          >
            Last reporting period for current CDC - Child Day Care - undefined funding
          </label>
          <span
            class="usa-hint text-italic"
          >
            Last reporting period is only required if this child's funding is no longer active.
          </span>
          <select
            aria-required="true"
            class="usa-select"
            id="last-reporting-period"
            name="last-reporting-period"
          >
            <option
              value=""
            >
              - Select -
            </option>
          </select>
        </div>
        <div
          class="display-flex flex-align-center"
        >
          <button
            class="usa-button usa-button--outline bx--btn bx--btn--primary"
            tabindex="0"
            type="button"
          >
            Cancel
          </button>
          <div
            class="margin-top-1"
          >
            <button
              class="usa-button bx--btn bx--btn--primary"
              tabindex="0"
              type="submit"
            >
              End current funding
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      class=""
    />
  </div>
  <div
    class="oec-card"
  >
    <div
      class="oec-card-cell"
    >
      <div
        class="display-flex flex-column flex-align-end"
      >
        <div>
          <button
            class="usa-button usa-button--unstyled margin-right-4 bx--btn bx--btn--primary"
            tabindex="0"
            type="button"
          >
            End current funding
          </button>
          <button
            class="usa-button usa-button--unstyled bx--btn bx--btn--primary"
            tabindex="0"
            type="button"
          >
            Change funding
          </button>
        </div>
      </div>
    </div>
    <div
      class=""
    />
  </div>
</DocumentFragment>
`;

exports[`EditRecord EnrollmentFundng ChangeFunding matches snapshot - without funding 1`] = `
<DocumentFragment>
  <div
    class="oec-card"
  >
    <div
      class="oec-card-cell"
    >
      <h3>
        Change funding
      </h3>
      <form
        class="usa-form usa-form"
        id="change-funding-form"
      >
        <fieldset
          aria-labelledby="fieldset-legend-funding-source-1"
          class="usa-fieldset"
          id="funding-source-1"
        >
          <legend
            id="fieldset-legend-funding-source-1"
          >
            <span
              class="usa-label"
            >
              Funding source options
            </span>
          </legend>
          <div />
        </fieldset>
        <div
          class="display-flex flex-align-center"
        >
          <button
            class="usa-button usa-button--outline bx--btn bx--btn--primary"
            tabindex="0"
            type="button"
          >
            Cancel
          </button>
          <div
            class="margin-top-1"
          >
            <button
              class="usa-button bx--btn bx--btn--primary"
              tabindex="0"
              type="submit"
            >
              Change current funding
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      class=""
    />
  </div>
  <div
    class="oec-card"
  >
    <div
      class="oec-card-cell"
    >
      <div
        class="display-flex flex-column flex-align-end"
      >
        <div>
          <button
            class="usa-button usa-button--unstyled bx--btn bx--btn--primary"
            tabindex="0"
            type="button"
          >
            Start new funding
          </button>
        </div>
      </div>
    </div>
    <div
      class=""
    />
  </div>
</DocumentFragment>
`;
